src/codecs/load_sndfile.c

```diff
@@ -0,0 +1,210 @@
+/*
+  SDL_mixer:  An audio mixer library based on the SDL library
+  Copyright (C) 1997-2023 Sam Lantinga <slouken@libsdl.org>
+
+  This software is provided 'as-is', without any express or implied
+  warranty.  In no event will the authors be held liable for any damages
+  arising from the use of this software.
+
+  Permission is granted to anyone to use this software for any purpose,
+  including commercial applications, and to alter it and redistribute it
+  freely, subject to the following restrictions:
+
+  1. The origin of this software must not be misrepresented; you must not
+     claim that you wrote the original software. If you use this software
+     in a product, an acknowledgment in the product documentation would be
+     appreciated but is not required.
+  2. Altered source versions must be plainly marked as such, and must not be
+     misrepresented as being the original software.
+  3. This notice may not be removed or altered from any source distribution.
+
+  This is the source needed to decode a file in any format supported by
+  libsndfile. The only externally-callable function is Mix_LoadSndFile_RW(),
+  which is meant to act as identically to SDL_LoadWAV_RW() as possible.
+
+  This file by Fabian Greffrath (fabian@greffrath.com).
+*/
+
+#include <SDL3/SDL_audio.h>
+
+#include "load_sndfile.h"
+
+#ifdef LOAD_SNDFILE
+
+#include <SDL3/SDL_loadso.h>
+
+#include <sndfile.h>
+
+static SNDFILE* (*SF_sf_open_virtual) (SF_VIRTUAL_IO *sfvirtual, int mode, SF_INFO *sfinfo, void *user_data);
+static int (*SF_sf_close) (SNDFILE *sndfile);
+static sf_count_t (*SF_sf_readf_short) (SNDFILE *sndfile, short *ptr, sf_count_t frames);
+static const char* (*SF_sf_strerror) (SNDFILE *sndfile);
+
+static int SNDFILE_init (void)
```

Comment:
There is no `SNDFILE_uninit` function. So libsndfile cannot be unloaded.

Reply:
It was never meant to be a music module. It is merely meant as a substitute for `SDL_LoadWAV_RW()` resp. an enhancement to `Mix_LoadWAV_RW()`, allowing the latter to load sound samples in some more exotic WAV-like formats. I guess the most common music formats are indeed already covered by SDL_mixer's modules.

Reply:
Ok, but even `Mix_Quit` will not unload libsndfile.
That's a bug imho.

Reply:
> Ok, but even `Mix_Quit` will not unload libsndfile. That's a bug imho.

@slouken Should I fix this?

Reply:
I can offer this:
```patch
diff --git a/src/codecs/load_sndfile.c b/src/codecs/load_sndfile.c
index 0b53d453..d4dfed2f 100644
--- a/src/codecs/load_sndfile.c
+++ b/src/codecs/load_sndfile.c
@@ -40,15 +40,14 @@ static int (*SF_sf_close) (SNDFILE *sndfile);
 static sf_count_t (*SF_sf_readf_short) (SNDFILE *sndfile, short *ptr, sf_count_t frames);
 static const char* (*SF_sf_strerror) (SNDFILE *sndfile);
 
+static int SNDFILE_loaded;
+static void *SNDFILE_lib;
+
 static int SNDFILE_init (void)
 {
-    static int SNDFILE_loaded;
-
     if (SNDFILE_loaded == 0)
     {
 #ifdef SNDFILE_DYNAMIC
-        static void *SNDFILE_lib;
-
         SNDFILE_lib = SDL_LoadObject(SNDFILE_DYNAMIC);
         if (SNDFILE_lib == NULL) {
             return -1;
@@ -80,6 +79,21 @@ static int SNDFILE_init (void)
     return 0;
 }
 
+void SNDFILE_uninit (void)
+{
+    if (SNDFILE_lib != NULL) {
+        SDL_UnloadObject(SNDFILE_lib);
+        SNDFILE_lib = NULL;
+
+        SF_sf_open_virtual = NULL;
+        SF_sf_close = NULL;
+        SF_sf_readf_short = NULL;
+        SF_sf_strerror = NULL;
+
+        SNDFILE_loaded = 0;
+    }
+}
+
 static sf_count_t sfvio_size(void *user_data)
 {
     SDL_RWops *RWops = user_data;
@@ -135,8 +149,10 @@ SDL_AudioSpec *Mix_LoadSndFile_RW (SDL_RWops *src, int freesrc,
     *audio_len = 0;
     SDL_memset(spec, 0, sizeof(*spec));
 
-    if (SNDFILE_init() != 0) {
-        goto done;
+    if (SNDFILE_loaded == 0) {
+        if (SNDFILE_init() != 0) {
+            goto done;
+        }
     }
 
     SDL_memset(&sfinfo, 0, sizeof(sfinfo));
@@ -205,6 +221,11 @@ SDL_AudioSpec *Mix_LoadSndFile_RW (SDL_RWops *src, int freesrc,
     return NULL;
 }
 
+void SNDFILE_uninit (void)
+{
+    // no-op
+}
+
 #endif
 
 /* vi: set ts=4 sw=4 expandtab: */
diff --git a/src/codecs/load_sndfile.h b/src/codecs/load_sndfile.h
index 6f5e6020..2f12e53e 100644
--- a/src/codecs/load_sndfile.h
+++ b/src/codecs/load_sndfile.h
@@ -34,6 +34,8 @@
 SDL_AudioSpec *Mix_LoadSndFile_RW (SDL_RWops *src, int freesrc,
         SDL_AudioSpec *spec, Uint8 **audio_buf, Uint32 *audio_len);
 
+void SNDFILE_uninit (void);
+
 #endif
 
 /* vi: set ts=4 sw=4 expandtab: */
diff --git a/src/mixer.c b/src/mixer.c
index 0af6ed7a..d11377f8 100644
--- a/src/mixer.c
+++ b/src/mixer.c
@@ -282,6 +282,7 @@ int Mix_Init(int flags)
 void Mix_Quit(void)
 {
     unload_music();
+    SNDFILE_uninit();
 }
 
 static int _Mix_remove_all_effects(int channel, effect_info **e);
```